### ISSUE TITLE
Select terms in predict.gamlss2

### ISSUE THREAD

**Issue by brute4c** [gamlss-dev/gamlss2#9]
When I make a mixed model with `gamlss2` I would like to only look at the contribution of the fixed effects of the model. To do this I would like to use `predict.gamlss2`with only selected fixed terms, but there seems to be an issue in doing this. 

When I use this approach with `gamlss2` I get different results compared to `mgcv` - see example below:  

```
## Example similar to: https://gamlss-dev.r-universe.dev/gamlss2/doc/manual.html#random  


## Data 
data("Ovary", package = "nlme")

## Model fit
fml         = follicles ~    Time  + ti(Mare, bs = "re") + ti(Mare, Time, bs = c("re", "cr"), k = c(11, 5))  ## smooth random effects
fit_gamlss2 = gamlss2(fml, data = Ovary)
fit_mgcv    = mgcv::gam(fml, data = Ovary) 

## Make predictions 
df_pred = 
data.frame(
  Mare = 2 |> factor(levels = levels(Ovary$Mare)), 
  Time = seq(0,1.2, by = 0.1)
) %>% 
  mutate(
    #mgcv 
    #mgcv_all    = fit_mgcv |> predict(newdata = df_new, type = 'response'),
    mgcv_ln      = fit_mgcv |> predict(newdata = df_new, type = 'response', terms = c('Time')),
    #mgcv_re     = fit_mgcv |> predict(newdata = df_new, type = 'response', terms = c('ti(Mare)')),
    #mgcv_ac     = fit_mgcv |> predict(newdata = df_new, type = 'response', terms = c('ti(Mare, Time')), 
    ##..gamlss2  
    #gamlss2_all = fit_gamlss2 |> predict(newdata = df_new, type = 'response'),
    gamlss2_ln   = fit_gamlss2 |> predict(newdata = df_new, type = 'response', terms = c('Time')),
    #gamlss2_re  = fit_gamlss2 |> predict(newdata = df_new, type = 'response', terms = c('ti(Mare)')),
    #gamlss2_ac  = fit_gamlss2 |> predict(newdata = df_new, type = 'response', terms = c('ti(Mare, Time')) 
  ) 

## Plot   
df_pred |> 
  pivot_longer(cols = starts_with( c('mgcv','gamlss2') ), names_to = 'fit_model', values_to = 'estimate') |>
  ggplot( aes(x = Time, y = estimate, color = fit_model, group = interaction(Mare, fit_model) )  )  + 
  geom_line() +     
  facet_wrap(~Mare)
```

**Comment by freezenik** [comment]
Dear Mikis,

it works, the problem is that you added a space " " in the character vector that specifies the terms you want to use for prediction. I have to test, if it is better to remove spaces internally.

Here is some updated code:

library("gamlss2")

## data 
data("Ovary", package = "nlme")

## formula
f <- follicles ~ s(Time)  + ti(Mare, bs = "re") + ti(Mare, Time, bs = c("re", "cr"), k = c(11, 5))

## fit model
b <- gamlss2(f, data = Ovary)

## predict only for Time.
p <- predict(b, model = "mu",
  terms = c("(Intercept)", "s(Time)"))

i <- order(Ovary$Time)
plot(p[i] ~ Ovary$Time[i], type = "l")

## predict with smooth random effects
p2 <- predict(b, model = "mu",
  terms = c("(Intercept)", "s(Time)", "ti(Mare)", "ti(Mare,Time)"))

## plot main effect with smooth random effects
plot(follicles ~ Time, data = Ovary, col = as.integer(Ovary$Mare))
lines(p[i] ~ Ovary$Time[i], lwd = 4)
for(j in 1:nlevels(Ovary$Mare)) {
  k <- Ovary$Mare == levels(Ovary$Mare)[j]
  x <- Ovary$Time[k]
  y <- p2[k]
  ii <- order(x)
  lines(y ~ x, col = j)
}

**Comment by brute4c** [comment]
Dear Nikolaus, 

Thank you for noticing my mistake - this would have answers my follow up question. 
My initial problem was slightly different and did not involve any cross terms (or spaces) so I think its still unanswered.  Please let me rephrase it, and notice I use `Time` in the equation and _not_ `s(Time)`. 
Also notice if you change `Time` to `s(Time)` everything works fine.


```
library("gamlss2")

## data 
data("Ovary", package = "nlme")

## formula 
f <- follicles ~   Time + ti(Mare, bs = "re") + ti(Mare, Time, bs = c("re", "cr"), k = c(11, 5))

## fit model
b <- gamlss2(f, data = Ovary)

# Predictions 
Ovary$fit_Time           = predict(b, model = "mu", terms = c("Time"))
Ovary$fit_all_exept_Time = predict(b, model = "mu", terms = c("(Intercept)"           , "ti(Mare)", "ti(Mare,Time)")) 
Ovary$fit_all            = predict(b, model = "mu", terms = c("(Intercept)", "s(Time)", "ti(Mare)", "ti(Mare,Time)")) 
Ovary$fit_all_sum        = Ovary$fit_all_exept_Time + Ovary$fit_Time
Ovary$fit_difference     = Ovary$fit_all - Ovary$fit_all_sum            # <---- This should be 0                                    
Ovary[1:10,]
```

**Comment by brute4c** [comment]
Thank you for the rapid replay. 

It seems to fix the problem, but still it seems to have some problems with interaction terms and it seems you do not select all terms by default. 


```{r}
library("gamlss2")
library("dplyr")

## data 
data("Ovary", package = "nlme")

# Add a baseline value for interaction term 
df_Ovary = 
  Ovary |> 
  as.data.frame() |> 
  group_by(Mare)  |> 
  mutate(follicles_0 = follicles[Time == min(Time)]) |> # initial number of follicles
  ungroup()


## formulas 
f1 <- follicles ~   follicles_0*Time                      + ti(Mare, bs = "re") + ti(Mare, Time, bs = c("re", "cr"), k = c(11, 5))
f2 <- follicles ~   follicles_0 + Time + follicles_0:Time + ti(Mare, bs = "re") + ti(Mare, Time, bs = c("re", "cr"), k = c(11, 5))
f3 <- follicles ~   follicles_0:Time + ti(Mare, bs = "re") + ti(Mare, Time, bs = c("re", "cr"), k = c(11, 5))


## fit model
b1 <- gamlss2(f1, data = df_Ovary)
b2 <- gamlss2(f2, data = df_Ovary)
b3 <- gamlss2(f3, data = df_Ovary)

# Issues 
predict(b1, model = "mu")                                                            # No terms selected? 
predict(b1, model = "mu",  type = 'response',  terms = c("follicles_0"), nogrep = T) # Should be constant per group? 
b2 |> summary()                                                                      # No interaction term - only main terms follicles_0 and Time 
b3 |> summary()                                                                      # Same as above   


```

**Comment by freezenik** [comment]
Thank you for catching this! I've just fixed it - could you please verify if everything is working as expected now?

**Comment by brute4c** [comment]
Thanks. 
This solved the first two issues, but not the last two - it seems to have a problem with the interaction term 'follicles_0:Time'. 

When using formula 'f1' in the previous example then you get the interaction term in the model and in the summary, but not when using f2 or f3. Also when using f3 you get the main terms which should not be included. 


```
b3 |> summary()  # no interactino term, but you get the main terms 
b3 |> predict(model = "mu",  type = 'response',  terms = c("follicles_0:Time"), nogrep = T) # only 0
```

**Comment by freezenik** [comment]
This took a bit longer, but I hope I fixed it for all cases now. Thanks for spotting this!

**Comment by brute4c** [comment]
Yes, you fixed it for all cases (except for spaces " " in the names that you mentioned at first, but this is not really a problem as long as you are aware of it).

Thank you very much for soving this issue.
